src/contract.jl

```diff
@@ -36,3 +40,66 @@ function contract_density_matrix(
   end
   return out
 end
+
+function ITensors.scalar(
+  alg::Algorithm, tn::Union{AbstractITensorNetwork,Vector{ITensor}}; kwargs...
+)
+  return contract(alg, tn; kwargs...)[]
+end
+
+function ITensors.scalar(
+  tn::Union{AbstractITensorNetwork,Vector{ITensor}}; alg="exact", kwargs...
+)
+  return scalar(Algorithm(alg), tn; kwargs...)
+end
+
+function logscalar(
+  tn::Union{AbstractITensorNetwork,Vector{ITensor}}; alg::String="exact", kwargs...
+)
+  return logscalar(Algorithm(alg), tn; kwargs...)
+end
+
+function logscalar(
+  alg::Algorithm"exact", tn::Union{AbstractITensorNetwork,Vector{ITensor}}; kwargs...
+)
+  s = scalar(alg, tn; kwargs...)
+  if s ≈ 0
+    tol = 1e-16
+    return -Inf
+  elseif isa(s, AbstractFloat) && s >= 0
+    return log(s)
+  else
+    return complex(s)
+  end
+end
+
+function logscalar(
+  alg::Algorithm,
+  tn::AbstractITensorNetwork;
+  (cache!)=nothing,
+  cache_construction_kwargs=default_cache_construction_kwargs(alg, tn),
+  update_cache=isnothing(cache!),
+  cache_update_kwargs=default_cache_update_kwargs(cache!),
+)
+  if isnothing(cache!)
+    cache! = Ref(cache(alg, tn; cache_construction_kwargs))
+  end
+
+  if update_cache
+    cache![] = update(cache![]; cache_update_kwargs...)
+  end
+
+  numerator_terms, denominator_terms = scalar_factors(cache![])
+  terms = vcat(numerator_terms, denominator_terms)
+  if any(≈(0), terms)
+    return -Inf
+  elseif all(t -> isa(t, AbstractFloat), terms) && all(>=(0), terms)
+    return sum(log.(numerator_terms)) - sum(log.((denominator_terms)))
+  else
+    return sum(log.(complex.(numerator_terms))) - sum(log.(complex.((denominator_terms))))
+  end
```

Comment:
This logic seems very complicated, I'm sure we can simplify it by thinking about it in a nicer what, factorizing into functions, etc.